### ISSUE TITLE
[Feature Request] Change of targetid-function to change observing Player/Entity

### ISSUE THREAD

**Issue by ZenBre4ker** [TTT-2/TTT2#668]
**Is your feature request related to a problem? Please describe.**
When I have an entity that I remotely control and use its view, I have to overwrite the current with my own targetid-Implementation to be able to see the target's ID. In my case this would be the supersheep by LeBroomer, I'm modding, but it could also be used for cameras like Cypher's  in Valorant and therefore be an interesting feature.

**Describe the solution you'd like**
An additional hook inside GM:HUDDrawTargetID() on line 236 in cl_targetid, where you can change the entity that is currently observing.

**Additional context**
I've already modded the code and will make a draft pull request. 
Some additional questions here are:
-Can you use :Eyepos() and :GetAimVector() on every entity?
-Should you then be able to see yourself?
-If you can see yourself, we would additionally have to change the filter for the trace and disable the ent==client check.
-As an extra it would be cool, if you could then outline yourself with green or in any special unused color

**Comment by TimGoll** [comment]
@lebroomer said something like this is already possible with the existing hooks, have you checked out this one:

```lua
---
-- Change the focused entity used for targetID.
-- @param Entity ent The focuces entity that should be changed
-- @param number distance The distance to the focused entity
-- @return Entity The new entity to replace the real one
-- @hook
-- @realm client
function GM:TTTModifyTargetedEntity(ent, distance)

end
```

**Comment by TimGoll** [comment]
After looking at your PR I see. LeBroomer suggested using my mentioned hook and shooting a new trace from within that hook. However I'm also open to the addition of a new hook.

**Comment by ZenBre4ker** [comment]
Well I was irritated for a moment but okay, that makes sense. That would mean double the calculation of the traceline and in that way, it would be possible.
With that additional hook you can just use all the code that is available and don't have to hack around the existing codebase.

**Comment by TimGoll** [comment]
I see the benefits of both approaches. If someone uses your addon, a new hook would greatly imnprove the performance. However to everyone not using your addon, this hook is a penalty on the performance. Drawhooks are always tricky and one has to outweigh the benefits and drawbacks.

Let's wait and see the opionions of @saibotk, @Alf21 and @Histalek

**Comment by ZenBre4ker** [comment]
Well in one hand, the performance drop would be only clientside, on the other hand I wonder if you couldnt set a convar if the hook should be called or not? So that it only affects those who use my addon. Or is a boolean/convar check every frame a huge drawback?

**Comment by TimGoll** [comment]
While a boolean check is faster, I don't think enabling a hook via a convar sounds like a good idea.

**Comment by lebroomer** [comment]
First of all, thank you for your contribution and the ideas. I think changing the behavior allowing to target yourself is a good idea considering addons (including my supersheep) where you are able to see yourself from a different perspective, some outline would be cool too. But I think the current hook is sufficient and there is no real need to add another hook to prevent a single trace from being shot. Additionally entities don't support ```GetAimVector()```.

**Comment by ZenBre4ker** [comment]
Well, if entities don't support this and this is why I asked, this solution doesnt make sense. My main reason isn't only the additional trace, but the fact, that your old had to be updated to work again or maybe I just did something wrong. But in general I think an addon developer should get a simple interface to use the TTT2 current targetid-implementation. 

Another idea would be to refactor the code into smaller snippets everyone could use. So instead of an additional hook just global functions inside huddrawtargetid that you can call with own parameters, so that you could replace huddrawtargetid with a simple copy and just other parameters.
But this still would lead to later update problems when you change those global functions..

So as I think this isnt good enough I would like to work on an improvement of this gm-function. But I'm unsure where to fix this and make it more general and useful for things like remote-controlled entities. This was just an idea, so if it's an interesting feature for you, I'd be open for other ideas to discuss.

**Comment by TimGoll** [comment]
I think you two misunderstood eachother. If xyou change the entity with a custom trace in the hook mentioned by me above, you can still use the targetID system without copypasting any code.

But lebroomers points seems important as well. How do you plan on handling entities without `GetAimVector()`?

**Comment by ZenBre4ker** [comment]
I asked for entities in the beginning of this thread, because I wasnt sure if they have the same functions like getaimvector. And broomer answered that they don't, so my solution doesnt work like this. So I'd need to have a hook that outputs the position and aimvector instead of using the player reference. which makes it way to specialised and is therefore not a good solution.

You both are right, that I can just shoot the trace myself by hooking to the modifyentity hook, which still is writing code twice, which I personally don't like and it still doesn't allow me to show myself on that remote camera. So additionally to write the trace code in the addon, I'd generally have to overwrite the rest too or just change the TTT2 function to a more general approach.

Hosting the functions of the raytrace and all the other functionalities in other global functions would allow me to reuse the code of ttt2 in any addon for example.

Or rewriting the code in general, so that it could allow for external cameras, maybe by adding some informations to the player in general.

Just some of many ways to include those functionalities in TTT2.

**Comment by TimGoll** [comment]
> and it still doesn't allow me to show myself on that remote camera

I really don't understand what you mean. You can show the targetID for each entity, even yourself.

But feel free to open a pullrequest with a suggestion of an external camera. I'm right now really busy with vgui and it is top priority for me. Probably for the rest of 2020 as well.

**Comment by ZenBre4ker** [comment]
I guess you are right. I mistook an "if ent== client then return" line for a definitive end. But as long as I change the changedEnt to the client, I can still draw the id for myself. 
I made a pullrequest, where you can see the comment on line 283, but I guess that's a mistake. I will look deeper into the existing hooks and maybe I can easily solve it. I just slowly begin to see an alternative with the way it is right now.

**Comment by ZenBre4ker** [comment]
Okay, I refactored it now instead, so everybody could use that function. That allowed my code to be shrunk down to 5 lines instead of copying the complete targetid-Implementation. With the current hook I can now easily change the entity that I found with the raycast, that I did with the refactored code. Remote cameras should now be able to easily show targetIds, without having to understand the raycasting.
